cpp/src/hash/hashing.cu

```diff
@@ -100,61 +100,84 @@ hash_partition_table(table_view const& input,
     // Compute which partition each row belongs to using bitwise partitioner
     auto partitioner = bitwise_partitioner<hash_value_type>{num_partitions};
     thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
-      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
-
+      hash_iterator + num_rows, row_partitions.begin(), partitioner);
   } else {
     // Compute which partition each row belongs to using modulo partitioner
     auto partitioner = modulo_partitioner<hash_value_type>{num_partitions};
     thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
-      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
+      hash_iterator + num_rows, row_partitions.begin(), partitioner);
   }
 
-  // Scatter input rows to output partitions given the partition map
-  auto const partition_map = column_view(data_type(INT32), num_rows,
-    row_partition_numbers.data().get());
-  auto output = experimental::detail::scatter_to_tables(input, partition_map, mr, stream);
-
-  // Pad with empty tables if we have less than num_partitions
-  if (output.size() < static_cast<size_t>(num_partitions)) {
-    std::generate_n(std::back_inserter(output), num_partitions - output.size(),
-      [&input]() { return experimental::empty_like(input); });
+  // Initialize gather maps and offsets to sequence
+  auto gather_map = rmm::device_vector<size_type>(num_rows);
+  auto d_offsets = rmm::device_vector<size_type>(num_rows);
+  thrust::sequence(rmm::exec_policy(stream)->on(stream),
+    gather_map.begin(), gather_map.end());
+  thrust::sequence(rmm::exec_policy(stream)->on(stream),
+    d_offsets.begin(), d_offsets.end());
+
+  // Sort sequence using partition map as key to generate gather maps
+  thrust::sort_by_key(rmm::exec_policy(stream)->on(stream),
+    row_partitions.begin(), row_partitions.end(), gather_map.begin());
+
+  // Reduce unique partitions to extract gather map offsets from sequence
+  auto end = thrust::unique_by_key(rmm::exec_policy(stream)->on(stream),
+    row_partitions.begin(), row_partitions.end(), d_offsets.begin());
+
+  auto output = experimental::detail::gather(input, gather_map.begin(), gather_map.end(),
+    false, false, false, mr, stream);
+
+  thrust::host_vector<size_type> unique_partitions(row_partitions.begin(), end.first);
+  thrust::host_vector<size_type> unique_offsets(d_offsets.begin(), end.second);
+
+  std::vector<size_type> offsets(num_partitions, input.num_rows());
+
+  // Pad offsets to insert empty partitions absent from unique_partitions
+  size_type next_partition = 0;
+  for (size_t i = 0; i < unique_partitions.size(); ++i) {
```

Comment:
I'm not quite following what's going on here.

Reply:
Because I use `unique` to reduce the sorted partition numbers, any empty partitions will be missing from the list. When building the offsets vector, I account for the empty partitions by duplicating the offset of the following non-empty partition.
An alternative approach could be to count the number of elements in each partition (including zeros for empty partitons) and then do an exclusive scan--I think that's what the previous implementation did.